doc/srcspan.md

```diff
@@ -0,0 +1,124 @@
+# Goal
+
+We want our debugger to provide user information at runtime about which
+location the currently-executing UPLC code was "sourced" from.
+Such a location is called a `SrcSpan` and is `(haskellHsFilename,startline,startcol,endline,endcol)`.
+
+There are two types of `SrcSpan` locations:
+
+## i. definition-site locations
+
+All GHC_Core (type) variables come by default annotated with a `SrcSpan` that points
+to where they are *defined* , in contrast to the current point that occur / are being called for.
+
+e.g.
+
+``` haskell
+f = 1 + 1
+
+main = f 3
+```
+
+The part of AST describing `f 3`, roughly in GHC_Core as  `(App (Var "f") (Lit 3))`,
+contains `SrcSpan` on the `Var "f"` node on where the `f` is defined, i.e. (Line 1, Col 1,...)
+
+## ii. call-site locations
+
+Function calls (applications) can optionally be annotated with a `SrcSpan` pointing to
+where they occur/ are called for (not where the function/arguments are defined).
+
+In the above example, the part of AST describing `f 3` --- `(App (Var "f") (Lit 3))` --- can be
+annotated with a SrcSpan of (Line 3, Col 8...).
+
+# How
+
+To achieve this we need:
+
+1. to collect the `SrcSpan` locations and
+2. propagate the `SrcSpans through our compilation artifacts (GHC_Core->PIR->PLC->UPLC).
+
+## 1. Collecting SrcSpans
+
+The definition-site locations are already provided by GHC and nothing extra needs to be done.
+The more interesting call-site locations need an extra step to instruct GHC to spit them out.
+
+There are (at least) 4 methods to make GHC annotate the GHC_Core with `SrcSpan`s,
+and hand it to us to feed it to our plutus-tx-plugin (start of our compiler pipeline):
+
+a. HPC (flag -fhpc)
+b. DWARF (flag -g )
+c. Prof (flag -prof)
+d. Ghci
+
+### a. HPC method
+
+The hpc method does not annotate the GHC core ast with precises `SrcSpan`s (the 5-tuple above) but
+with indices to a`SrcSpan` "array"" contained in a separately generated `mix` file.
+
+The good thing is that the resulting annotated GHC_core will end up being smaller in overall (serialised) size.
+The bad thing is that to find out the exact srcspan, we need an extra lookup at array "mix" file.
+
+### b. DWARF method
+
+Nothing peculiar to discuss about this method, and it looks like we already make use of it elsewhere in the plugin (for coverage):
```

Comment:
Do you know what kind of Tickish this gives us, is it `SourceNote`?

Reply:
Exactly